### PR TITLE
feat:[Pl-25315]: Introduced new Account Basic Role enabled through Feature Flag

### PR DESCRIPTION
**Pull request by Jimit09Gandhi**
## Describe your changes
Introduced new Account Basic Role enabled through Feature Flag.
Write Migration to add Account Basic as the Managed Role to all Users currently having Account Viewer Role.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/33844)

### CODE REVIEW

**Review by gupta-utkarsh — APPROVED** [review]
:lgtm:

> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33844#-:-N3fXhYcASXvGxPMfQxj:bnfp4nl)* status: :shipit: complete! ✓

___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/migration/AccountBasicRoleAssignmentAdditionMigration.java` line 34 at r1](https://reviewable.io/reviews/harness/harness-core/33844#-N3fWqxi16u4sLF0YGNB:-N3fWqxi16u4sLF0YGNC:bzay563) ([raw file](https://github.com/harness/harness-core/blob/7bb6d44c5c389567d166921b51d7bb6795afc1ff/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/migration/AccountBasicRoleAssignmentAdditionMigration.java#L34)):*
> ```Java
>     private final RoleAssignmentRepository roleAssignmentRepository;
>     private  final FeatureFlagHelperService featureFlagHelperService;
>     HashMap<String, Boolean> featureFlagForAccounts;
> ```

This should be private and final.
___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/migration/AccountBasicRoleAssignmentAdditionMigration.java` line 51 at r1](https://reviewable.io/reviews/harness/harness-core/33844#-N3fX0AXEvNe4_UyFLuy:-N3fX0AXEvNe4_UyFLuz:b-g7j564) ([raw file](https://github.com/harness/harness-core/blob/7bb6d44c5c389567d166921b51d7bb6795afc1ff/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/migration/AccountBasicRoleAssignmentAdditionMigration.java#L51)):*
> ```Java
>         do {
>             Pageable pageable = PageRequest.of(pageIndex, pageSize);
>             Criteria criteria = Criteria.where(RoleAssignmentDBOKeys.roleIdentifier).is("_account_viewer")
> ```

Please define "_account_viewer" it as constant in the class. Same goes for "_account_basic".

**Review by gupta-utkarsh — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33844)* status: 1 change requests. 0 of 1 approvals obtained. 

___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/migration/AccountBasicRoleAssignmentAdditionMigration.java` line 63 at r1](https://reviewable.io/reviews/harness/harness-core/33844#-N3fXtPO5LkbhkDtut16:-N3fXtPO5LkbhkDtut17:bnkpmvd) ([raw file](https://github.com/harness/harness-core/blob/7bb6d44c5c389567d166921b51d7bb6795afc1ff/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/migration/AccountBasicRoleAssignmentAdditionMigration.java#L63)):*
> ```Java
>                 String accountId = roleAssignment.getScopeIdentifier().replace("/ACCOUNT/", "");
>                 try {
>                     if (!featureFlagForAccounts.containsKey(accountId)) {
> ```

Please use computeIfAbsent function of hashMap for this.

File: access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/migration/AccountBasicRoleAssignmentAdditionMigration.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    public void migrate() {
+        log.info("AccountBasicRoleAssignmentAdditionMigration starts ...");
+        int pageSize = 1000;
+        int pageIndex = 0;
+
+        do {
+            Pageable pageable = PageRequest.of(pageIndex, pageSize);
+            Criteria criteria = Criteria.where(RoleAssignmentDBOKeys.roleIdentifier).is("_account_viewer")
+                    .and(RoleAssignmentDBOKeys.resourceGroupIdentifier).is(DEFAULT_ACCOUNT_LEVEL_RESOURCE_GROUP_IDENTIFIER)
+                    .and(RoleAssignmentDBOKeys.scopeLevel).is(HarnessScopeLevel.ACCOUNT.getName())
```
Comment (Jimit09Gandhi) [inline access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/migration/AccountBasicRoleAssignmentAdditionMigration.java:53]: Do we need to filter on Principal as User only?

**Review by Jimit09Gandhi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33844)* status: 3 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/migration/AccountBasicRoleAssignmentAdditionMigration.java` line 34 at r1](https://reviewable.io/reviews/harness/harness-core/33844#-N3fWqxi16u4sLF0YGNB:-N3p5TShBGqixJwlEW0y:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/7bb6d44c5c389567d166921b51d7bb6795afc1ff/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/migration/AccountBasicRoleAssignmentAdditionMigration.java#L34)):*
<details><summary><i>Previously, gupta-utkarsh (Utkarsh Gupta) wrote…</i></summary><blockquote>

This should be private and final.
</blockquote></details>

Done.
___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/migration/AccountBasicRoleAssignmentAdditionMigration.java` line 51 at r1](https://reviewable.io/reviews/harness/harness-core/33844#-N3fX0AXEvNe4_UyFLuy:-N3p5_K09D06mpjax9oK:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/7bb6d44c5c389567d166921b51d7bb6795afc1ff/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/migration/AccountBasicRoleAssignmentAdditionMigration.java#L51)):*
<details><summary><i>Previously, gupta-utkarsh (Utkarsh Gupta) wrote…</i></summary><blockquote>

Please define "_account_viewer" it as constant in the class. Same goes for "_account_basic".
</blockquote></details>

Done.
___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/migration/AccountBasicRoleAssignmentAdditionMigration.java` line 63 at r1](https://reviewable.io/reviews/harness/harness-core/33844#-N3fXtPO5LkbhkDtut16:-N3p5lFL2OPrYey59uwO:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/7bb6d44c5c389567d166921b51d7bb6795afc1ff/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/migration/AccountBasicRoleAssignmentAdditionMigration.java#L63)):*
<details><summary><i>Previously, gupta-utkarsh (Utkarsh Gupta) wrote…</i></summary><blockquote>

Please use computeIfAbsent function of hashMap for this.
</blockquote></details>

Done.

**Review by Jimit09Gandhi — COMMENTED** [review]
> Reviewed 2 of 5 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33844)* status: 3 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 1 approvals obtained.

**Review by gupta-utkarsh — DISMISSED** [review]
> Reviewed 2 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33844)* status: 1 file(s) to review. 

___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/migration/AccountBasicRoleAssignmentAdditionMigration.java` line 61 at r1](https://reviewable.io/reviews/harness/harness-core/33844#-N3ZXHBlj2gqDohiBcRL-r1-61:-N3qh1nL-9uwME-Zgqmj:bgs9pfd) ([raw file](https://github.com/harness/harness-core/blob/7bb6d44c5c389567d166921b51d7bb6795afc1ff/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/migration/AccountBasicRoleAssignmentAdditionMigration.java#L61)):*
<details><summary><i>Previously, Jimit09Gandhi (Jimit Gandhi) wrote…</i></summary><blockquote>

@gupta-utkarsh  Let me know better way to get Account Id from Role Assignment.
</blockquote></details>

Use scopeService.buildScopeFromScopeIdentifeir method. scope.getInstanceId will give you AccountId.

**Review by ankit-harness — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33844#-:-N3tiLh34g9J4ywD7Hac:bnfp4nl)* status: 1 file(s) to review.

**Review by reetika04 — APPROVED** [review]
> Reviewed 1 of 1 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33844)* status: 1 file(s) to review.

**Review by reetika04 — DISMISSED** [review]
> Reviewed 1 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33844)* status: :shipit: complete! ✓

**Comment by Jimit09Gandhi** [comment]
trigger ti4

**Comment by Jimit09Gandhi** [comment]
trigger ti2

**Comment by Jimit09Gandhi** [comment]
trigger ti1

**Comment by Jimit09Gandhi** [comment]
trigger tin4

**Comment by Jimit09Gandhi** [comment]
trigger tin4

**Comment by Jimit09Gandhi** [comment]
trigger ti4

**Comment by Jimit09Gandhi** [comment]
trigger ti2

**Comment by Jimit09Gandhi** [comment]
trigger ti0

**Comment by Jimit09Gandhi** [comment]
trigger ti1

**Comment by Jimit09Gandhi** [comment]
trigger ti2

**Comment by Jimit09Gandhi** [comment]
trigger ti4

**Review by gupta-utkarsh — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r4, 3 of 3 files at r5, 1 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33844#-:-N3xH1HVB7qPACHL4O0-:bnfp4nl)* status: :shipit: complete! ✓

**Pull request merged**